### PR TITLE
MF-1125 - add provision service docs

### PR DESCRIPTION
**Pull request by mteodor**
https://github.com/mainflux/mainflux/issues/1125

### CODE REVIEW

**Comment by drasko** [comment]
Please remove `WIP` label

File: docs/provision.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
+# Provision service
+
+Provision service provides an HTTP API to interact with [Mainflux][mainflux].
```
Comment (drasko) [inline docs/provision.md]: Tell us first what this service is, then what it provides.

File: docs/provision.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
+# Provision service
+
+Provision service provides an HTTP API to interact with [Mainflux][mainflux]. 
+Provision service is used to setup initial applications configuration i.e. things, channels, connections and certificates that will be required for the specific use case especially useful for gateway provision.  
+
+For gateways to communicate with [Mainflux][mainflux] configuration is required (mqtt host, thing, channels, certificates...). To get the configuration gateway will send a request to [Bootstrap][bootstrap] service providing `<external_id>` and `<external_key>` in request. To make a request to [Bootstrap][bootstrap] service you can use [Agent][agent] service on a gateway.
```
Comment (drasko) [inline docs/provision.md]: It is unclear what is `configuration gateway`. Yuo mean `unconfigured gateway` maybe?

File: docs/provision.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
+# Provision service
+
+Provision service provides an HTTP API to interact with [Mainflux][mainflux]. 
+Provision service is used to setup initial applications configuration i.e. things, channels, connections and certificates that will be required for the specific use case especially useful for gateway provision.  
+
+For gateways to communicate with [Mainflux][mainflux] configuration is required (mqtt host, thing, channels, certificates...). To get the configuration gateway will send a request to [Bootstrap][bootstrap] service providing `<external_id>` and `<external_key>` in request. To make a request to [Bootstrap][bootstrap] service you can use [Agent][agent] service on a gateway.  
+
+To create bootstrap configuration you can use [Bootstrap][bootstrap] or `Provision` service. [Mainflux UI][mfxui] uses [Bootstrap][bootstrap] service for creating gateway configurations.  `Provision` service should provide an easy way of provisioning your gateways i.e creating bootstrap configuration and as many things and channels that your setup requires.
```
Comment (drasko) [inline docs/provision.md]: This is a bit confusing. Why UI uses Bootstrap when there is Provisioning? Maybe this is temporary slution, and UI will move to Provision?

File: docs/provision.md (hunk 1 above)
Comment (drasko) [reply docs/provision.md]: Maybe als explain why we neded this service, the complexity of manual provisioning, what are the manual steps that can be automated, etc...

File: docs/provision.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+| MF_PROVISION_BS_SVC_URL             | Mainflux Bootstrap service URL                    | http://localhost/things/configs |
+| MF_PROVISION_BS_SVC_WHITELIST_URL   | Mainflux Bootstrap service whitelist URL          | http://localhost/things/state  |
+| MF_PROVISION_CERTS_SVC_URL          | Certificats service URL                           | http://localhost/certs         |
+| MF_PROVISION_X509_PROVISIONING      | Should X509 client cert be provisioned            | false                          |
+| MF_PROVISION_BS_CONFIG_PROVISIONING | Should thing config be saved in Bootstrap service | true                           |
+| MF_PROVISION_BS_AUTO_WHITEIST       | Should thing be auto whitelisted                  | true                           |
+| MF_PROVISION_BS_CONTENT             | Bootstrap service content                         | {}                             |
+
+By default, call to `/mapping` endpoint will create one thing and two channels (`control` and `data`) and connect it. If there is a requirement for different provision layout we can use [config](docker/configs/config.toml) file in addition to environment variables.
```
Comment (drasko) [inline docs/provision.md]: Take care with these links, they will not work in `docs` repo, as they are relative

**Review by drasko — CHANGES_REQUESTED** [review]

File: docs/provision.md (hunk 3 above)
Comment (mteodor) [reply docs/provision.md]: that is something that we need to consider

Comment (drasko) [reply docs/provision.md]: OK, once you reach the conclusion, please write it here as a note.

File: docs/provision.md [hunk 5]
```diff
@@ -1,13 +1,19 @@
 # Provision service
 
-Provision service provides an HTTP API to interact with [Mainflux][mainflux]. 
-Provision service is used to setup initial applications configuration i.e. things, channels, connections and certificates that will be required for the specific use case especially useful for gateway provision.  
+Provisioning is a process of configuration of an IoT platform in which system operator creates and sets-up different entities used in the platform - users, channels and things. It is part of process of setting up IoT application where
+we connect devices on edge with platform in cloud.
 
-For gateways to communicate with [Mainflux][mainflux] configuration is required (mqtt host, thing, channels, certificates...). To get the configuration gateway will send a request to [Bootstrap][bootstrap] service providing `<external_id>` and `<external_key>` in request. To make a request to [Bootstrap][bootstrap] service you can use [Agent][agent] service on a gateway.  
+For provisioning we can use [Mainflux CLI](./cli.md) creating users and for each node in the edge (eg. gateway) required number of things, channels, connecting them and creating certificates if needed.
```
Comment (manuio) [inline docs/provision.md]: I would use `https://github.com/mainflux/mainflux/tree/master/cli` here

**Review by manuio — CHANGES_REQUESTED** [review]

File: docs/provision.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+Provision service is used to setup initial applications configuration once user is created. Provision service creates  things, channels, connections and certificates. Once user is created we can use provision to create setup for edge node in one http request instead of issuing several cli commands. 
+
+Provision service provides an HTTP API to interact with [Mainflux][mainflux].  
+
+For gateways to communicate with [Mainflux][mainflux] configuration is required (mqtt host, thing, channels, certificates...). Gateway will send a request to [Bootstrap][bootstrap] service providing `<external_id>` and `<external_key>` in http request to get the configuration. To make a request to [Bootstrap][bootstrap] service you can use [Agent][agent] service on a gateway.  
+
+To create bootstrap configuration you can use [Bootstrap][bootstrap] or `Provision` service. [Mainflux UI][mfxui] uses [Bootstrap][bootstrap] service for creating gateway configurations.  `Provision` service should provide an easy way of provisioning your gateways i.e creating bootstrap configuration and as many things and channels that your setup requires.
```
Comment (manuio) [inline docs/provision.md]: Remove 1 white space before `  Provision service should provide`

**Review by manuio — CHANGES_REQUESTED** [review]

File: docs/provision.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+To create bootstrap configuration you can use [Bootstrap][bootstrap] or `Provision` service. [Mainflux UI][mfxui] uses [Bootstrap][bootstrap] service for creating gateway configurations.  `Provision` service should provide an easy way of provisioning your gateways i.e creating bootstrap configuration and as many things and channels that your setup requires.  
+
+Also you may use provision service to create certificates for each thing. Each service running on gateway may require more than one thing and channel for communication. Let's say that you are using services [Agent][agent] and [Export][exp] on a gateway you will need two channels for `Agent` (`data` and `control`) and one for `Export` and one thing. Additionally if you enabled mtls each service will need its own thing and certificate for access to [Mainflux][mainflux]. Your setup could require any number of things and channels this kind of setup we can call `provision layout`.
```
Comment (manuio) [inline docs/provision.md]: I'm not convinced about `Let's say that` in docs.
Also, is this correct? `you will need two channels for Agent (data and control) and one for Export and one thing.`

**Review by manuio — CHANGES_REQUESTED** [review]

File: docs/provision.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+By default, call to `/mapping` endpoint will create one thing and two channels (`control` and `data`) and connect it. If there is a requirement for different provision layout we can use [config](docker/configs/config.toml) file in addition to environment variables. 
+
+For the purposes of running provision as an add-on in docker composition environment variables seems more suitable. Environment variables are set in [.env][env].  
+
+Configuration can be specified in [config.toml][conftoml]. Config file can specify all the settings that environment variables can configure and in addition
+`/mapping` endpoint provision layout can be configured.
+
+In `config.toml` we can enlist array of things and channels that we want to create and make connections between them which we call provision layout.
+
+Metadata can be whatever suits your needs except that at least one thing needs to have `external_id` (which is populated with value from [request](#example)). Thing that has `external_id` will be used for creating bootstrap configuration which can be fetched with [Agent][agent].
```
Comment (manuio) [inline docs/provision.md]: `Things Metadata must have an "external_id" field which is populated with... `

**Review by manuio — CHANGES_REQUESTED** [review]

File: docs/provision.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
+# Provision service
+
+Provisioning is a process of configuration of an IoT platform in which system operator creates
+and sets-up different entities used in the platform - users, channels and things. 
+It is part of process of setting up IoT application where
+we connect devices on edge with platform in cloud.
+
+For provisioning we can use [Mainflux CLI][cli] for creating users
```
Comment (manuio) [inline docs/provision.md]: Is this return new line after `...creating users` ok?

**Review by manuio — CHANGES_REQUESTED** [review]

File: docs/provision.md (hunk 9 above)
Comment (mteodor) [reply docs/provision.md]: this new line will not be rendered, if that is what you mean

File: docs/provision.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+Configuration can be specified in [config.toml][conftoml]. 
+Config file can specify all the settings that environment variables can configure and in addition
+`/mapping` endpoint provision layout can be configured.
+
+In `config.toml` we can enlist array of things and channels that we want to create 
+and make connections between them which we call provision layout.
+
+Things Metadata can be whatever suits your needs. 
+Thing that has metadata with `external_id` will have bootstrap configuration created, `external_id` value will be populated with value from [request](#example)). 
+Bootstrap configuration which can be fetched with [Agent][agent].
```
Comment (dborovcanin) [inline docs/provision.md]: > Bootstrap configuration ~which~ can be fetched with [Agent][agent].

File: docs/provision.md [hunk 11]
```diff
@@ -0,0 +1,176 @@
+# Provision service
+
+Provisioning is a process of configuration of an IoT platform in which system operator creates
+and sets-up different entities used in the platform - users, channels and things.
```
Comment (dborovcanin) [inline docs/provision.md]: ..., and things.

File: docs/provision.md [hunk 12]
```diff
@@ -0,0 +1,176 @@
+# Provision service
+
+Provisioning is a process of configuration of an IoT platform in which system operator creates
+and sets-up different entities used in the platform - users, channels and things. 
+It is part of process of setting up IoT application where
```
Comment (dborovcanin) [inline docs/provision.md]: applications (plural)

File: docs/provision.md [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+we connect devices on edge with platform in cloud.
+
+For provisioning we can use [Mainflux CLI][cli] for creating users 
+and for each node in the edge (eg. gateway) required number of things, channels, connecting them and creating certificates if needed.
```
Comment (dborovcanin) [inline docs/provision.md]: ..., and creating certificates...

File: docs/provision.md [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+Provision service is used to setup initial applications configuration once user is created.
```
Comment (dborovcanin) [inline docs/provision.md]: ... is used to set up initial application...

File: docs/provision.md [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+It is part of process of setting up IoT application where
+we connect devices on edge with platform in cloud.
+
+For provisioning we can use [Mainflux CLI][cli] for creating users 
+and for each node in the edge (eg. gateway) required number of things, channels, connecting them and creating certificates if needed.
+
+Provision service is used to setup initial applications configuration once user is created. 
+Provision service creates  things, channels, connections and certificates. 
+Once user is created we can use provision to create setup for edge node 
+in one http request instead of issuing several cli commands.
```
Comment (dborovcanin) [inline docs/provision.md]: Please use capital letters in acronyms: HTTP request, MQTT protocol, CLI (commands), TLS, API... Check for these in the rest of the document, as well.

File: docs/provision.md [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+Config file can specify all the settings that environment variables can configure and in addition
+`/mapping` endpoint provision layout can be configured.
+
+In `config.toml` we can enlist array of things and channels that we want to create 
+and make connections between them which we call provision layout.
+
+Things Metadata can be whatever suits your needs. 
+Thing that has metadata with `external_id` will have bootstrap configuration created, `external_id` value will be populated with value from [request](#example)). 
+Bootstrap configuration which can be fetched with [Agent][agent].
+For channels metadata `type` is reserved for `control` and `data` which we use with [Agent][agent].
```
Comment (dborovcanin) [inline docs/provision.md]: For channels, metadata...

File: docs/provision.md [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+
+To create bootstrap configuration you can use [Bootstrap][bootstrap] or `Provision` service. 
+[Mainflux UI][mfxui] uses [Bootstrap][bootstrap] service for creating gateway configurations. 
+`Provision` service should provide an easy way of provisioning your gateways 
+i.e creating bootstrap configuration and as many things and channels that your setup requires.  
+
+Also you may use provision service to create certificates for each thing. 
+Each service running on gateway may require more than one thing and channel for communication. 
+If, for example, you are using services [Agent][agent] and [Export][exp] on a gateway you will need two channels for `Agent` (`data` and `control`) and one thing for `Export`. 
+Additionally if you enabled mtls each service will need its own thing and certificate for access to [Mainflux][mainflux].
```
Comment (dborovcanin) [inline docs/provision.md]: Additionally, if you enable mTLS...

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: docs/platform-management.md [hunk 18]
```diff
@@ -0,0 +1,324 @@
+Provisioning is a process of configuration of an IoT platform in which system operator creates and sets-up different entities
+used in the platform - users, channels and things.
```
Comment (drasko) [inline docs/platform-management.md]: No need to break line here

File: docs/platform-management.md [hunk 19]
```diff
@@ -0,0 +1,324 @@
...
+
+Before proceeding, make sure that you have created a new account, and obtained
+an authorization key.
+
+### Provisioning things
+
+> This endpoint will be depreciated in 0.11.0.  It will be replaced with the bulk endpoint currently found at /things/bulk.
+
+Things are created by executing request `POST /things` with a JSON payload.
+Note that you will also need `user_auth_token` in order to create things
```
Comment (drasko) [inline docs/platform-management.md]: You can move this to the upper line

File: docs/platform-management.md [hunk 20]
```diff
@@ -0,0 +1,324 @@
...
+```
+{
+      "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE1MjMzODg0NzcsImlhdCI6MTUyMzM1MjQ3NywiaXNzIjoibWFpbmZsdXgiLCJzdWIiOiJqb2huLmRvZUBlbWFpbC5jb20ifQ.cygz9zoqD7Rd8f88hpQNilTCAS1DrLLgLg4PRcH-iAI"
+}
+```
+
+## System provisioning
+
+Before proceeding, make sure that you have created a new account, and obtained
+an authorization key.
```
Comment (drasko) [inline docs/platform-management.md]: No need for line break here

File: docs/platform-management.md [hunk 21]
```diff
@@ -0,0 +1,324 @@
...
+HTTP/1.1 201 Created
+Content-Type: application/json
+Location: /channels/19daa7a8-a489-4571-8714-ef1a214ed914
+Date: Tue, 10 Apr 2018 11:30:07 GMT
+Content-Length: 0
+```
+
+### Bulk provisioning channels
+
+Multiple channels can be created by executing a `POST /things/bulk` request with a JSON payload.  The payload should contain a JSON array of the channels to be created.  If there is an error any of the channels, none of the channels will be created.
```
Comment (drasko) [inline docs/platform-management.md]: This line is actually good. Renders nicely, easy to read. Maybe align all to this style.

**Review by drasko — CHANGES_REQUESTED** [review]

**Review by drasko — DISMISSED** [review]
LGTM

**Review by blokovi — DISMISSED** [review]

File: docs/platform-management.md [hunk 22]
```diff
@@ -1,5 +1,4 @@
-Provisioning is a process of configuration of an IoT platform in which system operator creates and sets-up different entities
-used in the platform - users, channels and things.
+Provisioning is a process of configuration of an IoT platform in which system operator creates and sets-up different entities used in the platform - users, channels and things.
```
Comment (dborovcanin) [inline docs/platform-management.md]: , and things.

File: docs/platform-management.md [hunk 23]
```diff
@@ -116,29 +107,25 @@ Content-Length: 1105
 }
 ```
 
-You can specify `offset` and `limit` parameters in order to fetch specific
-group of things. In that case, your request should look like:
+You can specify `offset` and `limit` parameters in order to fetch specific group of things. In that case, your request should look like:
```
Comment (dborovcanin) [inline docs/platform-management.md]: ...to fetch a specific group...

File: docs/platform-management.md [hunk 24]
```diff
@@ -116,29 +107,25 @@ Content-Length: 1105
...
 
-```
+```bash
 curl -s -S -i --cacert docker/ssl/certs/ca.crt -H "Authorization: <user_auth_token>" https://localhost/things?offset=0&limit=5
 ```
 
-You can specify `name` and/or `metadata` parameters in order to fetch specific
-group of things. When specifiying metadata you can specify just a part of the metadata json you want to match
+You can specify `name` and/or `metadata` parameters in order to fetch specific group of things. When specifiying metadata you can specify just a part of the metadata json you want to match
```
Comment (dborovcanin) [inline docs/platform-management.md]: specifying, JSON

File: docs/provision.md [hunk 25]
```diff
@@ -0,0 +1,156 @@
+# Provision service
+
+Provisioning is a process of configuration of an IoT platform in which system operator creates and sets-up different entities used in the platform - users, channels, and things. It is part of process of setting up IoT applications where we connect devices on edge with platform in cloud.
+
+For provisioning we can use [Mainflux CLI][cli] for creating users and for each node in the edge (eg. gateway) required number of things, channels, connecting them, and creating certificates if needed.
+
+Provision service is used to setup initial application configuration once user is created. Provision service creates  things, channels, connections, and certificates. Once user is created we can use provision to create setup for edge node in one HTTP request instead of issuing several CLI commands.
```
Comment (dborovcanin) [inline docs/provision.md]: Provision service is used to set up...
...to create a setup...

File: docs/provision.md [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+
+Provision service provides an HTTP API to interact with [Mainflux][mainflux].  
+
+For gateways to communicate with [Mainflux][mainflux] configuration is required (MQTT host, thing, channels, certificates...). Gateway will send a request to [Bootstrap][bootstrap] service providing `<external_id>` and `<external_key>` in HTTP request to get the configuration. To make a request to [Bootstrap][bootstrap] service you can use [Agent][agent] service on a gateway.  
+
+To create bootstrap configuration you can use [Bootstrap][bootstrap] or `Provision` service. [Mainflux UI][mfxui] uses [Bootstrap][bootstrap] service for creating gateway configurations. `Provision` service should provide an easy way of provisioning your gateways i.e creating bootstrap configuration and as many things and channels that your setup requires.  
+
+Also you may use provision service to create certificates for each thing. Each service running on gateway may require more than one thing and channel for communication.
```
Comment (dborovcanin) [inline docs/provision.md]: Also, you...

File: docs/provision.md [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+The service is configured using the environment variables presented in the following [table][config]. Note that any unset variables will be replaced with their default values.
+
+
+By default, call to `/mapping` endpoint will create one thing and two channels (`control` and `data`) and connect it. If there is a requirement for different provision layout we can use [config][conftoml] file in addition to environment variables. 
+
+For the purposes of running provision as an add-on in docker composition environment variables seems more suitable. Environment variables are set in [.env][env].  
+
+Configuration can be specified in [config.toml][conftoml]. Config file can specify all the settings that environment variables can configure and in addition `/mapping` endpoint provision layout can be configured.
+
+In `config.toml` we can enlist array of things and channels that we want to create and make connections between them which we call provision layout.
```
Comment (dborovcanin) [inline docs/provision.md]: ... enlist an array...

File: docs/provision.md [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+
+  [channels.metadata]
+    type = "data"
+```
+
+## Authentication
+In order to create necessary entities provision service needs to authenticate against Mainflux. 
+To provide authentication credentials to the provision service you can pass it in as an environment variable or in a config file as Mainflux user and password or as API token (that can be issued on `/users` or `/keys` endpoint of [authn][authn]. 
+
+Additionally users or API token can be passed in Authorization header, this authentication takes precedence over others.
```
Comment (dborovcanin) [inline docs/provision.md]: Additionally, ...

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: docs/platform-management.md (hunk 24 above)
Comment (manuio) [reply docs/platform-management.md]: Also, after match `.`

File: docs/platform-management.md (hunk 23 above)
Comment (mteodor) [reply docs/platform-management.md]: i think that this is ok

File: docs/platform-management.md (hunk 22 above)
Comment (mteodor) [reply docs/platform-management.md]: note that this `,` is optional

Comment (drasko) [reply docs/platform-management.md]: What's this comment about? Putting `,` before `and` is gramatically incorrect.

Comment (mteodor) [reply docs/platform-management.md]: I think it is not

**Review by manuio — APPROVED** [review]
LGTM

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**